### PR TITLE
ref(types): Add types to the pipeline module

### PR DESCRIPTION
**Pull request by mgaeta**
This PR adds types to `src/sentry/pipeline/`.

### CODE REVIEW

File: src/sentry/integrations/base.py [hunk 1]
```diff
@@ -123,11 +123,6 @@ class is just a descriptor for how that object functions, and what behavior
     it provides (such as extensions provided).
     """
 
-    # a unique identifier (e.g. 'slack').
-    # Used to lookup sibling classes and the ``key`` used when creating
-    # Integration objects.
-    key: Optional[str] = None
```
Comment (evanh) [inline src/sentry/integrations/base.py:129]: Why did you remove this?

Comment (wedamija) [reply src/sentry/integrations/base.py:129]: I think this will cause an error if a subclass hasn't implemented `key`. If it's required we might want to make it an abstract property.

File: src/sentry/pipeline/base.py [hunk 2]
```diff
@@ -41,13 +47,13 @@ class Pipeline:
     using the ``update_config`` method.
     """
 
-    pipeline_name = None
-    provider_manager = None
-    provider_model_cls = None
+    pipeline_name: str
+    provider_manager: Any
```
Comment (wedamija) [inline src/sentry/pipeline/base.py:51]: Could be worth defining some kind of interface here at some point, not necessarily this pr though. Seems like we just require that the manager implements a `.get` that returns a `PipelineProvider`

**Review by wedamija — COMMENTED** [review]
lgtm, just not sure about the `key` removal

**Comment by RyanSkonnord** [comment]
Agree with @wedamija [about](https://github.com/getsentry/sentry/pull/32863/files#r833502916) removing the `key` attribute, otherwise LGTM.

File: src/sentry/pipeline/base.py [hunk 3]
```diff
@@ -41,13 +47,13 @@ class Pipeline:
...
+    provider_model_cls: Model
```
Comment (RyanSkonnord) [inline src/sentry/pipeline/base.py:52]: Should this be `Type[Model]`?

File: src/sentry/pipeline/provider.py [hunk 4]
```diff
@@ -1,29 +1,43 @@
...
+
+class PipelineProvider(abc.ABC):
     """
     A class implementing the PipelineProvider interface provides the pipeline
     views that the Pipeline will traverse through.
     """
 
-    def __init__(self):
-        self.config = {}
+    #: A unique identifier (e.g. 'slack'). Used to lookup sibling classes and
```
Comment (untitaker) [inline src/sentry/pipeline/provider.py:16]: I suspect this was accidentally moved in the wrong class

File: src/sentry/integrations/base.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/integrations/base.py:129]: I tried moving `key` to `PipelineProvider` as an abstract class property so that `IntegrationProvider` and others could use it without having to redefine it. 
Screen Shot 2022-04-04 at 4 52 38 PM
I'm going to revert because I think a refactor like that should happen in python 3.10 when we have a better decorator.

Comment (mgaeta) [reply src/sentry/integrations/base.py:129]: It turns out that I actually really do have to add `key` and `name` to `PipelineProvider` so I've pulled that into a separate PR: https://github.com/getsentry/sentry/pull/33348.

**Comment by mgaeta** [comment]
Ok this should now be totally separated from the Provider key stuff.

File: src/sentry/pipeline/__init__.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
-from .base import Pipeline
-from .provider import PipelineProvider
+from .provider import PipelineProvider  # isort:skip
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/pipeline/__init__.py:1]: In `src/sentry/integrations/base.py`, we do 
```py
from sentry.pipeline import PipelineProvider
```
which goes here and imports these submodules in order. But because `src/sentry/pipeline/base.py` does 
```py
from . import PipelineProvider
```
There would be a circular dependency.

**Review by NisanthanNanthakumar — APPROVED** [review]
